### DISCUSSION TITLE
Pretty printing relations

### DISCUSSION THREAD

**Discussion by ydewit** [souffle-lang/souffle#2368] [General]
I am wondering if there is a technique that other people have used to pretty print relations.

Ideally, I would like to see a simple template language that could generate syntax text out of AST relations (think: Source -> AST -> Facts -> Outputs -> Annotated AST). But I am sure that is asking too much.

I guess using the API to traverse through the output facts?

**Comment by quentin** [comment]
Can you give an example of pretty-printing you would like to have? Is this for the stdout output?

You might want to write and register your own `WriteStreamFactory` in `souffle::IOSystem`, and then `.output relation(IO="registered_name")`. Although currently that is only possible with a compiled souffle program that you instanciate manually from your own c++ code.

Note that you need set `--output-dir` to anything but `-` otherwise the default stdout printer is selected regardless of the `IO=` value.

We could probably extend souffle to be able to call arbitrary write stream factories from a dynamically linked library (same way as functors).
